### PR TITLE
Add Reverse Proxy

### PR DESCRIPTION
**Pull request by xyliew25**
**Implementation Details**
- Check commit messages

### CODE REVIEW

File: frontend/src/configs.js [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
-export const URL_COMM_SVC_CONNECT = URI_COMM_SVC;
+// Communication service
+const PREFIX_COMM_SVC = '/api/communication-service';
+const URL_COMM_SVC = URI_SVC + PREFIX_COMM_SVC;
+const PREFIX_COMM_SVC_READ_CHAT = '/read';
 export const URL_COMM_SVC_READ_CHAT = URL_COMM_SVC + PREFIX_COMM_SVC_READ_CHAT;
+export const URL_COMM_SVC_CONNECT = URI_SVC; // Socket
+
+// Collaboration service
+export const URL_COLLAB_SVC_CONNECT = URI_SVC; // Socket
```
Comment (DesmondTo) [inline frontend/src/configs.js:34]: I do not see the endpoints for question service, can you help to add here (if any) for consistency? Thanks!

File: frontend/src/context/user-context.js [hunk 2]
```diff
@@ -2,18 +2,11 @@ import { createContext, useState } from 'react';
 
 import axios from 'axios';
 
-import { URL_USER_SVC } from '../configs';
+import { 
+  URL_USER_SVC_SIGNIN, URL_USER_SVC_JWT_VERIFICATION, URL_USER_SVC_LOGOUT, URL_USER_SVC_UPDATE, URL_USER_SVC_DELETE
```
Comment (DesmondTo) [inline frontend/src/context/user-context.js:6]: Perhaps can style this vertically so looks better

File: frontend/src/pages/Match/index.js [hunk 3]
```diff
@@ -33,29 +33,25 @@ function MatchPage() {
   const alertCtx = useContext(AlertContext);
   const userCtx = useContext(UserContext);
   // Difficulty
-  const [difficulty, setDifficulty] = useState('');
+  const difficulty = location.state?.difficulty;
```
Comment (DesmondTo) [inline frontend/src/pages/Match/index.js:36]: Good practice for avoiding redundant usage of hook 👍

File: frontend/src/pages/Match/index.js [hunk 4]
```diff
@@ -33,29 +33,25 @@ function MatchPage() {
...
   // Socket
-  const [socket, setSocket] = useState(null);
+  const socket = io(URL_MATCHING_SVC_CONNECT, {
+    path: '/api/matching-service/socket',
```
Comment (DesmondTo) [inline frontend/src/pages/Match/index.js:39]: Please put this hardcoded path into config.js, thanks! Let's consolidate all the URL into the file. 😄

File: frontend/src/pages/Match/index.js [hunk 5]
```diff
@@ -33,29 +33,25 @@ function MatchPage() {
...
   useEffect(() => {
-    if (location.state?.difficulty) {
-      setDifficulty(location.state.difficulty);
-    } else {
+    if (!difficulty) {
       navigate('/home');
       alertCtx.onShow('Please select a difficulty level!');
     }
-  }, [location.state, alertCtx, navigate]);
+  }, [difficulty, alertCtx, navigate]);
```
Comment (DesmondTo) [inline frontend/src/pages/Match/index.js:53]: have you check if these dependency array will cause infinite component rendering using `console.log`? If done, please ignore this.

File: frontend/src/pages/Match/index.js [hunk 6]
```diff
@@ -33,29 +33,25 @@ function MatchPage() {
...
   // Socket
-  useEffect(() => {
-    setSocket(io.connect(URL_MATCHING_SVC));
-  }, []);
   useEffect(() => {
```
Comment (DesmondTo) [inline frontend/src/pages/Match/index.js:55]: There are many useEffect hook in this file, can you help to split them with empty line and perhaps add some comment on what they do? thanks!

File: frontend/src/pages/Room/index.js [hunk 7]
```diff
@@ -13,11 +13,14 @@ import QuestionBox from '../../components/ui/question/QuestionBox';
...
 import styles from './Room.module.css';
 
-import { URL_COMM_SVC_CONNECT } from '../../configs';
-import { URL_COLLAB_SVC } from '../../configs';
+import { URL_COMM_SVC_CONNECT, URL_COLLAB_SVC_CONNECT } from '../../configs';
 
-const comm_socket = io(URL_COMM_SVC_CONNECT);
-const collab_socket = io(URL_COLLAB_SVC);
+const comm_socket = io(URL_COMM_SVC_CONNECT, {
+  path: '/api/communication-service/socket',
```
Comment (DesmondTo) [inline frontend/src/pages/Room/index.js:19]: Please put this hardcoded path into config.js, thanks! Let's consolidate all the URL into the file. 😄

File: frontend/src/pages/Room/index.js [hunk 8]
```diff
@@ -13,11 +13,14 @@ import QuestionBox from '../../components/ui/question/QuestionBox';
...
+});
+const collab_socket = io(URL_COLLAB_SVC_CONNECT, {
+  path: '/api/collaboration-service/socket',
```
Comment (DesmondTo) [inline frontend/src/pages/Room/index.js:22]: Please put this hardcoded path into config.js, thanks! Let's consolidate all the URL into the file. 😄

File: nginx/default.conf [hunk 9]
```diff
@@ -0,0 +1,36 @@
+server {
```
Comment (DesmondTo) [inline nginx/default.conf:1]: Perhaps can call this folder as `gateway` instead of generic `nginx`

File: nginx/default.conf [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  listen 8080;
+
+  location /api/user-service {
+    proxy_pass http://user-service:8000/api/user;
+  }
+
+  location /api/matching-service/socket {
+    proxy_pass http://matching-service:8001;
+    proxy_http_version 1.1;
```
Comment (DesmondTo) [inline nginx/default.conf:10]: Perhaps can add comment to explain these upgrade and 1.1 keepalive connection is for enabling WebSockets? Thanks 😄 Same for those below.

File: nginx/default.conf [hunk 11]
```diff
@@ -0,0 +1,36 @@
+server {
+  listen 8080;
+
+  location /api/user-service {
+    proxy_pass http://user-service:8000/api/user;
```
Comment (DesmondTo) [inline nginx/default.conf:5]: Have you tested if the JWT token can be sent through this reverse proxy for authentication? If it works, ignore this comment.

**Review by DesmondTo — CHANGES_REQUESTED** [review]
LGTM, but there are some nitty gritty stuff need to be addressed.

File: frontend/src/pages/Match/index.js (hunk 4 above)
Comment (xyliew25) [reply frontend/src/pages/Match/index.js:39]: Agree

File: nginx/default.conf (hunk 9 above)
Comment (xyliew25) [reply nginx/default.conf:1]: Sure

File: frontend/src/configs.js [hunk 12]
```diff
@@ -1,17 +1,34 @@
...
+// Matching service
+export const URL_MATCHING_SVC_CONNECT = URI_SVC; // Socket
 
-export const URL_COMM_SVC_CONNECT = URI_COMM_SVC;
+// Communication service
+const PREFIX_COMM_SVC = '/api/communication-service';
+const URL_COMM_SVC = URI_SVC + PREFIX_COMM_SVC;
+const PREFIX_COMM_SVC_READ_CHAT = '/read';
 export const URL_COMM_SVC_READ_CHAT = URL_COMM_SVC + PREFIX_COMM_SVC_READ_CHAT;
+export const URL_COMM_SVC_CONNECT = URI_SVC; // Socket
```
[the remark this answers is not in the document]
Comment (xyliew25) [reply frontend/src/configs.js:31]: Sure

File: frontend/src/configs.js (hunk 1 above)
Comment (xyliew25) [reply frontend/src/configs.js:34]: Sure, I can help to setup the base URL first, which can then be further extended in [this PR](https://github.com/CS3219-AY2223S1/cs3219-project-ay2223s1-g28/pull/50).

File: nginx/default.conf (hunk 10 above)
Comment (xyliew25) [reply nginx/default.conf:10]: Agree

File: frontend/src/configs.js [hunk 13]
```diff
@@ -1,17 +1,34 @@
...
+export const URL_USER_SVC_DELETE = URL_USER_SVC + PREFIX_USER_SVC_DELETE;
+
+// Matching service
+export const URL_MATCHING_SVC_CONNECT = URI_SVC; // Socket
 
-export const URL_COMM_SVC_CONNECT = URI_COMM_SVC;
+// Communication service
+const PREFIX_COMM_SVC = '/api/communication-service';
+const URL_COMM_SVC = URI_SVC + PREFIX_COMM_SVC;
+const PREFIX_COMM_SVC_READ_CHAT = '/read';
```
[the remark this answers is not in the document]
Comment (xyliew25) [reply frontend/src/configs.js:29]: But these two `PREFIX`s, i.e., `PREFIX_COMM_SVC` and `PREFIX_COMM_SVC_READ_CHAT`, are not of the same kind. Something like user service above.

[the remark this answers is not in the document]
Comment (DesmondTo) [reply frontend/src/configs.js:29]: Then perhaps can go with 2

[the remark this answers is not in the document]
Comment (xyliew25) [reply frontend/src/configs.js:29]: Sure

File: frontend/src/pages/Match/index.js (hunk 6 above)
Comment (xyliew25) [reply frontend/src/pages/Match/index.js:55]: Sure

File: nginx/default.conf (hunk 11 above)
Comment (xyliew25) [reply nginx/default.conf]: Yup can

File: frontend/src/configs.js [hunk 14]
```diff
@@ -1,20 +1,49 @@
...
-const PREFIX_USER_SVC = '/api/user';
-const PREFIX_COMM_SVC = '/api/chat';
-const PREFIX_QUES_SVC = '/api/question';
-
-export const URL_USER_SVC = URI_USER_SVC + PREFIX_USER_SVC;
-export const URL_MATCHING_SVC = URI_MATCHING_SVC;
-export const URL_COMM_SVC = URI_COMM_SVC + PREFIX_COMM_SVC;
-export const URL_COLLAB_SVC = URI_COLLAB_SVC;
-export const URL_QUES_SVC = URI_QUES_SVC + PREFIX_QUES_SVC;
+/////////////// Reverse Proxy ///////////////
```
Comment (DesmondTo) [inline frontend/src/configs.js:1]: Very neat to split them like this. Good job!

**Review by DesmondTo — APPROVED** [review]
LGTM, good job!

**Pull request merged**